src/ReplicatedStorage/Client/Tool/Gun/init.lua

```diff
@@ -190,6 +225,14 @@ function Gun:OnActivated()
 	raycastParams.FilterType = Enum.RaycastFilterType.Blacklist
 	raycastParams.FilterDescendantsInstances = blacklist
 
+	local function rayCheck(originalPosition: Vector3, startPosition: Vector3, endPosition: Vector3): Vector3?
+		local rayCheckDirection = (endPosition - startPosition).Unit * range
+		local raycastResult = workspace:Raycast(startPosition, rayCheckDirection, raycastParams)
+		if raycastResult and (raycastResult.Position - originalPosition).Magnitude > 4 then
+			return raycastResult.Position
+		end
```

Comment:
What is this Magnitude > 4 check doing?

Reply:
I'm mostly confused why you do this more than anything. As far as I can tell, you are only using this for visually changing the position if it moves a decent bit, and that's it, right? Is that just for style, or, do you have some other reason?

Reply:
If it's just for style then it's probably fine.

Reply:
The reason for this is the gun's barrel can be of the other side of a wall/corner while the character is behind the cover. There used to be a strategy in Eclipsis (before they fixed it) where people would place Platform blocks or slab blocks, stand behind them and position their gun's barrel on the other side of the block. This allowed them to shoot people while being fully covered. This is a very unfun experience for the opponents. What this code does is that besides the primary ray that is fired from the barrel to the mouse position it also fires 2 secondary rays from the head and the shoulder. Then it checks if the position of those rayResults are within 1 stud (in the previous code 4 studs) of the primary ray's position. If both the head's ray and the right shoulder's ray are blocked by something I assume the character is behind cover and I do not allow the bullet to go past the cover, instead I use the position of the rightShoulder's ray.

I hope this makes sense

Reply:
Seems to make sense yeah. Alternatively, you could raycast once, and on hit, raycast in the reverse direction back to the source and make sure nothing is hit. If something is hit, you could probably use that position instead. This should work in all cases where it makes sense (even if all ray sources are inside the part geometry for example) and would probably be simpler.

Reply:
Right, that makes sense. I applied the new changes